### ISSUE TITLE
Segmenting and saving each class as image to allow OCR

### ISSUE THREAD

**Issue by elnazsn1988** [hpanwar08/detectron2#9]
## 🚀 Feature
Hi, is there an internal feature which lets each classed be saved as a seperate segment, or image? I am trying to identify tables, seperate and then run through a tabular data analyzer and ocr - so far am able to get the image predictions with your code, but not the actual annotations/segmented fields for further analysis/ocr.

## Pitch

Note that you can implement many features by [extending detectron2](https://detectron2.readthedocs.io/tutorials/extend.html).
See [projects](https://github.com/facebookresearch/detectron2/tree/master/projects/) for some examples.
We would only consider adding new features if they are relevant to many users.

**Comment by hpanwar08** [comment]
@elnazsn1988 You can try the below code. It extracts the bounding boxes and then crop the image based on the bounding box and save.

```python
classes = ['text', 'title', 'list', 'table', 'figure']
default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
img = detectron2.data.detection_utils.read_image(path_to_image, format="BGR")
predictions = default_predictor(img)
instances = predictions["instances"].to('cpu')

pred_classes = instances.pred_classes
labels = [classes[i] for i in pred_classes]
boxes = instances.pred_boxes
if isinstance(boxes, detectron2.structures.boxes.Boxes):
    boxes = boxes.tensor.numpy()
else:
    boxes = np.asarray(boxes)

for label, bbox in zip(labels, boxes):
    if label == "table":
        cropped_img = img.crop(bbox)
        croppped_img.save(f"{label}_{bbox}.png")
```

**Comment by elnazsn1988** [comment]
@hpanwar08 thanks for the above, it throws errors at img being extracted as a numpy object :     cropped_img = img.crop(bbox)
AttributeError: 'numpy.ndarray' object has no attribute 'crop'

**Comment by hpanwar08** [comment]
@elnazsn1988 I have updated the code, it should work now

**Comment by akshay94950** [comment]
where to add the code ?which file?

**Comment by hpanwar08** [comment]
@akshay94950 You can add this code in a new python file and run it.

**Comment by akshay94950** [comment]
@hpanwar08 
the above code is throwing error:
 49     if label == "text":
     50         cropped_img = img.crop(bbox)
---> 51         croppped_img.save(f"{label}_{bbox}.png")
     52 
     53 

NameError: name 'croppped_img' is not defined

**Comment by hpanwar08** [comment]
can you show the entire code.

**Comment by akshay94950** [comment]
import argparse
import glob
import multiprocessing as mp
import os
import time
import cv2
import tqdm
import detectron2
from detectron2.utils.logger import setup_logger
setup_logger()
from detectron2.config import get_cfg
from detectron2.data.detection_utils import read_image
from detectron2.utils.logger import setup_logger
from detectron2.engine import DefaultPredictor
from detectron2.config import get_cfg
from detectron2.utils.visualizer import Visualizer
from detectron2.data import MetadataCatalog
from detectron2.engine import DefaultTrainer
from detectron2.config import get_cfg

from predictor import VisualizationDemo

# constants
WINDOW_NAME = "COCO detections"
classes = ['text', 'title', 'list', 'table', 'figure']
cfg = get_cfg()
cfg.merge_from_file("/content/gdrive/My Drive/detectron2-master/configs/DLA_mask_rcnn_R_101_FPN_3x.yaml")
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.2  # set threshold for this model
cfg.MODEL.WEIGHTS = os.path.join( '/content/gdrive/My Drive/Unt', 'model_final.pth')
cfg.MODEL.DEVICE='cpu'
default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
img = detectron2.data.detection_utils.read_image("/content/gdrive/My Drive/unt/page1.jpg" , format="BGR")
predictions = default_predictor(img)
instances = predictions["instances"].to('cpu')

pred_classes = instances.pred_classes
labels = [classes[i] for i in pred_classes]
boxes = instances.pred_boxes
if isinstance(boxes, detectron2.structures.boxes.Boxes):
    boxes = boxes.tensor.numpy()
else:
    boxes = np.asarray(boxes)
print('b',boxes)
from PIL import Image
img = Image.fromarray(img)


for label, bbox in zip(labels, boxes):
    if label == "text":
        cropped = img.crop(bbox)
        print("l",cropped)
        croppped.save("l.jpg")

**Comment by akshay94950** [comment]
it  was because of spelling mistake in that above code

**Comment by akshay94950** [comment]
thanks bro its working.....@hpanwar08

**Comment by elnazsn1988** [comment]
> @elnazsn1988 I have updated the code, it should work now

HI Hpanwar - codes working great, do you happen to know how I can annotate a new image and retrain the existing weights for the new image? Ive been trying to donwload the full dataset from:
wget -O datz/publaynet.tar.gz https://dax-cdn.cdn.appdomain.cloud/dax-publaynet/1.0.0/publaynet.tar.gz
but somehow the download is unstable so I cant get the whole dataset to retrain and annotate/add my file.

**Comment by hpanwar08** [comment]
I did not understand your question. If you want to annotate new images then you can have a look at https://github.com/wkentaro/labelme
It's better to download the dataset in parts train-0.tar.gz, train-1.tar.gz etc rather than as one huge file.

**Comment by elnazsn1988** [comment]
> I did not understand your question. If you want to annotate new images then you can have a look at https://github.com/wkentaro/labelme
> It's better to download the dataset in parts train-0.tar.gz, train-1.tar.gz etc rather than as one huge file.

@hpanwar08 is there a way to add a new annotated image and train your preteained model on it without retraining the whole thing? As in use your weights from the trained model and retrain on a new image set.

**Comment by hpanwar08** [comment]
Yes, you can annotated your custom dataset and save it in COCO format and train using train_net_dla.py

**Comment by akshay94950** [comment]
can we obtain predictions output image without figures class for using it in ocr?

**Comment by akshay94950** [comment]
![image](https://user-images.githubusercontent.com/62750721/78504844-1c78e300-778d-11ea-981b-1127c39071be.png)

**Comment by jocelynguo** [comment]
> @elnazsn1988 You can try the below code. It extracts the bounding boxes and then crop the image based on the bounding box and save.
> 
> ```python
> classes = ['text', 'title', 'list', 'table', 'figure']
> default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
> img = detectron2.data.detection_utils.read_image(path_to_image, format="BGR")
> predictions = default_predictor(img)
> instances = predictions["instances"].to('cpu')
> 
> pred_classes = instances.pred_classes
> labels = [classes[i] for i in pred_classes]
> boxes = instances.pred_boxes
> if isinstance(boxes, detectron2.structures.boxes.Boxes):
>     boxes = boxes.tensor.numpy()
> else:
>     boxes = np.asarray(boxes)
> 
> from PIL import Image
> img = Image.fromarray(img)
> 
> for label, bbox in zip(labels, boxes):
>     if label == "table":
>         cropped_img = img.crop(bbox)
>         croppped_img.save(f"{label}_{bbox}.png")
> ```

**One follow-up question: How to extract the confident score from each of the boxes?**

**Comment by hpanwar08** [comment]
> 
> 
> can we obtain predictions output image without figures class for using it in ocr?or how to use it for text extraction ?

You could directly crop the images with bounding box

**Comment by hpanwar08** [comment]
> 
> 
> > @elnazsn1988 You can try the below code. It extracts the bounding boxes and then crop the image based on the bounding box and save.
> > ```python
> > classes = ['text', 'title', 'list', 'table', 'figure']
> > default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
> > img = detectron2.data.detection_utils.read_image(path_to_image, format="BGR")
> > predictions = default_predictor(img)
> > instances = predictions["instances"].to('cpu')
> > 
> > pred_classes = instances.pred_classes
> > labels = [classes[i] for i in pred_classes]
> > boxes = instances.pred_boxes
> > if isinstance(boxes, detectron2.structures.boxes.Boxes):
> >     boxes = boxes.tensor.numpy()
> > else:
> >     boxes = np.asarray(boxes)
> > 
> > from PIL import Image
> > img = Image.fromarray(img)
> > 
> > for label, bbox in zip(labels, boxes):
> >     if label == "table":
> >         cropped_img = img.crop(bbox)
> >         croppped_img.save(f"{label}_{bbox}.png")
> > ```
> 
> **One follow-up question: How to extract the confident score from each of the boxes?**

`instances.scores` will give you the predicted probabilities for each instance.

**Comment by jocelynguo** [comment]
> > > @elnazsn1988 You can try the below code. It extracts the bounding boxes and then crop the image based on the bounding box and save.
> > > ```python
> > > classes = ['text', 'title', 'list', 'table', 'figure']
> > > default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
> > > img = detectron2.data.detection_utils.read_image(path_to_image, format="BGR")
> > > predictions = default_predictor(img)
> > > instances = predictions["instances"].to('cpu')
> > > 
> > > pred_classes = instances.pred_classes
> > > labels = [classes[i] for i in pred_classes]
> > > boxes = instances.pred_boxes
> > > if isinstance(boxes, detectron2.structures.boxes.Boxes):
> > >     boxes = boxes.tensor.numpy()
> > > else:
> > >     boxes = np.asarray(boxes)
> > > 
> > > from PIL import Image
> > > img = Image.fromarray(img)
> > > 
> > > for label, bbox in zip(labels, boxes):
> > >     if label == "table":
> > >         cropped_img = img.crop(bbox)
> > >         croppped_img.save(f"{label}_{bbox}.png")
> > > ```
> > 
> > 
> > **One follow-up question: How to extract the confident score from each of the boxes?**
> 
> `instances.scores` will give you the predicted probabilities for each instance.

Thank you!

**Comment by akshay94950** [comment]
@hpanwar08 how to get the images saved using above code in order,which can be later used for an ocr using tesseract?

**Comment by hpanwar08** [comment]
These images will be saved in the same directory in which you run the code.

**Comment by akshay94950** [comment]
I mean the images are not in order. I want to use it for OCR. How to sort
in order and get of the original input?

**Comment by hpanwar08** [comment]
what do you want to sort on?

**Comment by akshay94950** [comment]
The images are saved in the order of predictions but if I use the images in
that order I can't extract meaningful text from the input image. I want all
classes of predictions except figures and I want to use it for a text to
speech module.

**Comment by hpanwar08** [comment]
exclude the figure label when iterating the labels
`if label != "figure":
    ...
`

**Comment by hpanwar08** [comment]
Now I got what you are saying. 
You may need sort segments based on the bounding box location. But it would depend on the page layout. Some images may have two column layout, some may have one, some may have mixed layout. You need write your logic based on these constraints.

**Comment by akshay94950** [comment]
Yeah... That's matter. Can you help something regarding this?thanks for the
help you shown till now..

**Comment by hpanwar08** [comment]
One of the solution could be, write a classifier to classify the type of image e.g. single column, 2 column etc. Then write logic based on the type of the image e.g sorting based on x or y.

**Comment by akshay94950** [comment]
can you please help me writing the logic of a single column image

**Comment by hpanwar08** [comment]
try this

```python
classes = ['text', 'title', 'list', 'table', 'figure']
default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
img = detectron2.data.detection_utils.read_image(path_to_image, format="BGR")
predictions = default_predictor(img)
instances = predictions["instances"].to('cpu')

pred_classes = instances.pred_classes
labels = [classes[i] for i in pred_classes]
boxes = instances.pred_boxes
if isinstance(boxes, detectron2.structures.boxes.Boxes):
    boxes = boxes.tensor.numpy()
else:
    boxes = np.asarray(boxes)

from PIL import Image
img = Image.fromarray(img)

# bbox = [xmin. ymin, xmax, ymax], 
# sort by bboxes and lebels by ymin
sorted_by_bbox_ymin = sorted(zip(boxes, labels), key=lambda x: x[0][1])
boxes, labels = list(*sorted_by_bbox_ymin)
boxes = list(boxes)
labels = list(labels)

for idx, (label, bbox) in enumerate(zip(labels, boxes)):
    if label == "text":
        cropped_img = img.crop(bbox)
        croppped_img.save(f"{idx}_{label}_{bbox}.png")
```

**Comment by deeptigoyal** [comment]
Hi, I also want to run ocr on text label images obtained after running detectron, whrein order of images is required to obtain meaningful text.

I ran the above sorted code and ended up getting error:

**list expected at most 1 arguments, got 18**

If anyone has been able to crack it, then please help.
![error2](https://user-images.githubusercontent.com/13421799/80457209-7600a780-891e-11ea-8a99-ba524042968c.png)

**Comment by hpanwar08** [comment]
Fixed the code, should work now.

**Comment by deeptigoyal** [comment]
Yooo...**thanks** 

:)

**Comment by akshay94950** [comment]
Thanks.... For this work... It's working nice to some of images...

**Comment by akshay94950** [comment]
Thanks.... @hpanwar08  For this work... It's working nice to some of images...'did you use the complete publaynet dataset for training?if not i like to train the remaining data in your model for that can you specify the portion of dataset you doesnt use ?

**Comment by tnav96** [comment]
Hi, I am wondering if you have an example python script which I could run to test the pretrained detectron? That would be a hughe help

**Comment by hpanwar08** [comment]
There is a command written in the README.md which you can try.

**Comment by tnav96** [comment]
I was more looking for something like this which I could also manipulate afterwards. I am working in jupyter lab. Do you also have a simple example with such a structure but also including the implementation of cfg and all the required libraries?

> try this
> 
> ```python
> classes = ['text', 'title', 'list', 'table', 'figure']
> default_predictor = detectron2.engine.defaults.DefaultPredictor(cfg)
> img = detectron2.data.detection_utils.read_image(path_to_image, format="BGR")
> predictions = default_predictor(img)
> instances = predictions["instances"].to('cpu')
> 
> pred_classes = instances.pred_classes
> labels = [classes[i] for i in pred_classes]
> boxes = instances.pred_boxes
> if isinstance(boxes, detectron2.structures.boxes.Boxes):
>     boxes = boxes.tensor.numpy()
> else:
>     boxes = np.asarray(boxes)
> 
> from PIL import Image
> img = Image.fromarray(img)
> 
> # bbox = [xmin. ymin, xmax, ymax], 
> # sort by bboxes and lebels by ymin
> sorted_by_bbox_ymin = sorted(zip(boxes, labels), key=lambda x: x[0][1])
> boxes, labels = list(zip(*sorted_by_bbox_ymin))
> boxes = list(boxes)
> labels = list(labels)
> 
> for idx, (label, bbox) in enumerate(zip(labels, boxes)):
>     if label == "text":
>         cropped_img = img.crop(bbox)
>         croppped_img.save(f"{idx}_{label}_{bbox}.png")
> ```

**Comment by hpanwar08** [comment]
You could use the above code, it should get the predictions for you. You need to install detectron2 first and copy the cfg from this repo to your installation.

**Comment by hpanwar08** [comment]
> Thanks.... @hpanwar08 For this work... It's working nice to some of images...'did you use the complete publaynet dataset for training?if not i like to train the remaining data in your model for that can you specify the portion of dataset you doesnt use ?

I used almost half of the publaynet data, it is mentioned in the README file.

**Comment by tnav96** [comment]
I installed detectron2 already but if I just run the code above like that it obviously says that the library is not defined and cfg is not defined. I guess there is missing something like:

import argparse
import glob
import multiprocessing as mp
import os
import time
import cv2
import tqdm
import detectron2
from detectron2.utils.logger import setup_logger
setup_logger()
from detectron2.config import get_cfg
from detectron2.data.detection_utils import read_image
from detectron2.utils.logger import setup_logger
from detectron2.engine import DefaultPredictor
from detectron2.config import get_cfg
from detectron2.utils.visualizer import Visualizer
from detectron2.data import MetadataCatalog
from detectron2.engine import DefaultTrainer
from detectron2.config import get_cfg
#

cfg = get_cfg()
cfg.merge_from_file("detectron2-master/configs/DLA_mask_rcnn_X_101_32x8d_FPN_3x.yaml")
cfg.DATALOADER.NUM_WORKERS = 2
cfg.MODEL.WEIGHTS = os.path.join( 'trained', 'model_final_trimmed.pth')
cfg.MODEL.DEVICE='cpu'
